### PR TITLE
Fixed #34226 -- Fixed QuerySet.select_related() with multiple FilteredRelations to the OneToOneField.

### PR DESCRIPTION
**Pull request by zhu**
https://code.djangoproject.com/ticket/34226

### CODE REVIEW

File: tests/known_related_objects/tests.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from django.test import TestCase
 
-from .models import Organiser, Pool, PoolStyle, Tournament
+from django.db.models import FilteredRelation
+from .models import Organiser, Pool, PoolStyle, Tournament, AnotherPoolStyle
```
Comment (jarek-pacocha) [inline tests/known_related_objects/tests.py:4]:
```suggestion

from .models import AnotherPoolStyle, Organiser, Pool, PoolStyle, Tournament
```

File: tests/known_related_objects/tests.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from django.test import TestCase
```
Comment (jarek-pacocha) [inline tests/known_related_objects/tests.py:1]:
```suggestion
from django.db.models import FilteredRelation
from django.test import TestCase
```

**Review by jarek-pacocha — COMMENTED** [review]
I focused only on the check errors here since you discussed technicalities in a [#34226 ticket](https://code.djangoproject.com/ticket/34226)

Please double check me.

**Comment by charettes** [comment]
I haven't looked in details into the origin of the issue but the fact `f` is used by the other loop and inner loop 

https://github.com/django/django/blob/160d4e5d08e63c7848279d99e2e6558f712ca66c/django/db/models/sql/compiler.py#L1156

https://github.com/django/django/blob/160d4e5d08e63c7848279d99e2e6558f712ca66c/django/db/models/sql/compiler.py#L1217

Seems like it could be causing all sorts of weird things depending on whether `related_fields` is empty or not.

This was not an issue before e1ae2b00504ba30481285b2bd767d1ad561bf4be as `f` was always overridden by the `self.query.setup_joins([name], opts, root_alias)` call.

I'd suggest that the inner loop usage of `f` be renamed to something else and that we use `partial` here.

**Comment by felixxm** [comment]
> I haven't looked in details into the origin of the issue but the fact `f` is used by the other loop and inner loop
> 
> https://github.com/django/django/blob/160d4e5d08e63c7848279d99e2e6558f712ca66c/django/db/models/sql/compiler.py#L1156
> 
> https://github.com/django/django/blob/160d4e5d08e63c7848279d99e2e6558f712ca66c/django/db/models/sql/compiler.py#L1217
> 
> Seems like it could be causing all sorts of weird things depending on whether `related_fields` is empty or not.
> 
> This was not an issue before [e1ae2b0](https://github.com/django/django/commit/e1ae2b00504ba30481285b2bd767d1ad561bf4be) as `f` was always overridden by the `self.query.setup_joins([name], opts, root_alias)` call.
> 
> I'd suggest that the inner loop usage of `f` be renamed to something else and that we use `partial` here.

Good catch :+1: I've prepared PR, see #16402.

**Review by felixxm — COMMENTED** [review]
@zhu Thanks :+1: Welcome aboard :boat:

**Pull request merged**